homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -0,0 +1,99 @@
+"""Config flow to configure Xiaomi Miio."""
+import logging
+
+from miio import DeviceException, gateway
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_TOKEN
+
+# pylint: disable=unused-import
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_GATEWAY = "gateway"
+DEFAULT_GATEWAY_NAME = "Xiaomi Gateway"
+
+GATEWAY_CONFIG = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_TOKEN): vol.All(str, vol.Length(min=32, max=32)),
+        vol.Optional(CONF_NAME, default=DEFAULT_GATEWAY_NAME): str,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema({vol.Optional(CONF_GATEWAY, default=False): bool})
+
+
+class XiaomiMiioFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Xiaomi Miio config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the Xiaomi Miio flow."""
+        self._gateway = None
+        self._gateway_info = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # Check which device needs to be connected.
+            if user_input.get(CONF_GATEWAY):
+                return self.async_show_form(
```

Comment:
When we want to move to the next step, await and return that step, don't call show form.